### PR TITLE
US24 Created GlobalExceptionHandler, Exceptions hierarchy

### PR DESCRIPTION
**Pull request by AndriiHorbenko**

### CODE REVIEW

File: social_network_api/src/main/java/com/danit/socialnetwork/exception/GlobalExceptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    return new ResponseEntity<>(error, HttpStatus.NOT_FOUND);
+  }
+
+  @ExceptionHandler
+  public ResponseEntity<ErrorResponse> catchHeaderPhotoNotFoundException(HeaderPhotoNotFoundException exception) {
+    ErrorResponse error = new ErrorResponse();
+    error.setStatus(HttpStatus.NOT_FOUND.value());
+    error.setMessage(exception.getMessage());
+    error.setTimeStamp(System.currentTimeMillis());
+    return new ResponseEntity<>(error, HttpStatus.NOT_FOUND);
```
Comment (miegami) [inline social_network_api/src/main/java/com/danit/socialnetwork/exception/GlobalExceptionHandler.java:39]: You can create one method for all these exceptions

File: social_network_api/src/main/java/com/danit/socialnetwork/service/UserService.java [hunk 2]
```diff
@@ -3,6 +3,7 @@
 import com.danit.socialnetwork.model.DbUser;
 
 import java.io.IOException;
+import java.util.Optional;
```
Comment (miegami) [inline social_network_api/src/main/java/com/danit/socialnetwork/service/UserService.java:6]: Do you use it?

File: social_network_api/src/main/java/com/danit/socialnetwork/service/UserServiceImpl.java [hunk 3]
```diff
@@ -8,30 +11,63 @@
...
 public class UserServiceImpl implements UserService {
 
   private final UserRepository userRepository;
 
+  private <T extends Object> Boolean isEmpty(T temp) {
+    Optional<T> maybeExist = Optional.ofNullable(temp);
+    if (maybeExist.isEmpty()) {
+      return true;
+    } else {
+      return false;
```
Comment (miegami) [inline social_network_api/src/main/java/com/danit/socialnetwork/service/UserServiceImpl.java:28]: return maybeExist.isEmpty()

File: social_network_api/src/main/java/com/danit/socialnetwork/service/UserServiceImpl.java [hunk 4]
```diff
@@ -8,30 +11,63 @@
...
+
   @Override
   public DbUser findByUsername(String username) {
-    return userRepository.findByUsername(username);
+    DbUser maybeUser = userRepository.findByUsername(username);
+    if (isEmpty(maybeUser)) {
+      throw new UserNotFoundException("User with username " + username + " not found");
+    } else {
+      return maybeUser;
+    }
```
Comment (miegami) [inline social_network_api/src/main/java/com/danit/socialnetwork/service/UserServiceImpl.java:40]: You can remove else, just return

File: social_network_api/src/main/java/com/danit/socialnetwork/exception/GlobalExceptionHandler.java (hunk 1 above)
Comment (AndriiHorbenko) [reply social_network_api/src/main/java/com/danit/socialnetwork/exception/GlobalExceptionHandler.java:39]: @miegami I thought about this I can do to add just one AppUserError. If I just return informaton it is ok, but for the future if somebody would like to process some of them separately it better to have them separately.

Comment (AndriiHorbenko) [reply social_network_api/src/main/java/com/danit/socialnetwork/exception/GlobalExceptionHandler.java:20]: Done, if somebody would need this he would add ExceptionHandler for definite Exception

Comment (miegami) [reply social_network_api/src/main/java/com/danit/socialnetwork/exception/GlobalExceptionHandler.java:20]: I mean that you can write something like @ExceptionHandler({FirstException.class, SecondException.class})

**Review by miegami — APPROVED** [review]

**Review by AlexKhmarenko — APPROVED** [review]

**Pull request merged**